### PR TITLE
GPU Array blog post

### PR DESCRIPTION
**Pull request by SimonDanisch**

### CODE REVIEW

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+of [Transpiler.jl](https://github.com/SimonDanisch/Transpiler.jl) for an OpenGL and OpenCL backend and uses [CUDAnative](https://github.com/JuliaGPU/CUDAnative.jl/) for nvidia based hardware.
+It also offers a threaded Julia backend which can even utilize Xeon Phi's and is great for debugging the GPU kernels.
+GPUArrays also offers abstractions over the intrinsics of the different backends and makes it possible to write GPU programs
+completely in Julia for all hardware.
+
+CuArrays is very similar to GPUArrays cudanative backend and was first supposed to be an extension of that,
+but we decided to concentrate on CUDA support in it's own package to create a much leaner package.
+This is a fair approach for now and we plan to merge things back at some point to decrease fragmentation of the ecosystem.
+
+Like CuArrays, GPUArrays supports functions like [broadcast](https://julialang.org/blog/2017/01/moredots), `mapreduce`, indexing, `vcat`/`hcat` and all functionality that comes for free from the support of these generic functions, e.g. sum, maximum, etc.
```
Comment (ararslan) [inline blog/_posts/2017-8-25-gpuarrays.md]: The code formatting applied to function names is inconsistent here. I'd add backticks to broadcast, sum, and maximum.

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+It can run the whole suite and print out reports automatically. We're planning to integrate it with [Julia's speed center](http://kristofferc.bitnamiapp.com/), to have
+an interactive website where one can explore the results. Until then, the static report [over at GPUBenchmarks.jl](https://github.com/JuliaGPU/GPUBenchmarks.jl/blob/master/results/results.md) has to do the trick.
+This is the first release of GPUBenchmarks, so please treat the numbers with care!
+
+For the future, we plan to continue integrate existing libraries like [CUDNN](https://github.com/JuliaGPU/CUDNN.jl) and [CUSolver](https://github.com/JuliaGPU/CUSOLVER.jl) into GPUArrays as we head for feature parity with Julia's Base Array type.
+We also hope to make the ecosystem much more stable and improve the abstractions needed to write platform independent GPU code.
+
+All in all, Julia can now finally offer single source GPU programming, much like what [SyCL](https://www.google.de/search?q=SyCL&oq=SyCL&aqs=chrome..69i57j0l5.193j0j4&sourceid=chrome&ie=UTF-8) has promised for years.
```
Comment (ararslan) [inline blog/_posts/2017-8-25-gpuarrays.md]: Did you mean to link to a German Google search here?

**Comment by ararslan** [comment]
This is really cool!

File: blog/_posts/2017-8-25-gpuarrays.md (hunk 1 above)
Comment (SimonDanisch) [reply blog/_posts/2017-8-25-gpuarrays.md]: thanks!

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+but we decided to concentrate on CUDA support in it's own package to create a much leaner package.
+This is a fair approach for now and we plan to merge things back at some point to decrease fragmentation of the ecosystem.
+
+Like CuArrays, GPUArrays supports functions like [broadcast](https://julialang.org/blog/2017/01/moredots), `mapreduce`, indexing, `vcat`/`hcat` and all the functionality that comes for free from the support of these generic functions, e.g. sum, maximum, etc.
+
+On top of that GPUArrays wraps CU/CL-BLAS and CU/CL-FFT enabling a wide range of applications.
+
+To get an idea of where we are in terms of performance, I started writing a benchmark suite for Julia's GPU array packages including [ArrayFire](https://github.com/gaika/ArrayFire.jl).
+It can run the whole suite and print out reports automatically. We're planning to integrate it with [Julia's speed center](http://kristofferc.bitnamiapp.com/), to have an interactive website where one can explore the results. Until then, the static report [over at GPUBenchmarks.jl](https://github.com/JuliaGPU/GPUBenchmarks.jl/blob/master/results/results.md) has to do the trick.
```
Comment (KristofferC) [inline blog/_posts/2017-8-25-gpuarrays.md]: I'm not really comfortable linking to the Julia speed center thing until more work has been done on it, and a bunch of things fixed.

Comment (SimonDanisch) [reply blog/_posts/2017-8-25-gpuarrays.md]: ah, sorry didn't realize. Thanks for the heads up!

File: blog/_posts/2017-8-25-gpuarrays.md (hunk 1 above)
Comment (ararslan) [reply blog/_posts/2017-8-25-gpuarrays.md]: Once that's addressed I think this is good to go!

**Comment by SimonDanisch** [comment]
cc @MikeInnes @maleadt

**Comment by ViralBShah** [comment]
I think it would greatlt help to have the benchmarks included here in the post.

https://github.com/JuliaGPU/GPUBenchmarks.jl/blob/master/results/results.md

File: blog/_posts/2017-8-25-gpuarrays.md (hunk 1 above)
Comment (SimonDanisch) [reply blog/_posts/2017-8-25-gpuarrays.md]: oh i misunderstood and thought you corrected it already^^

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+---
+layout: post
+title: "Julia's GPU ecosystem: A status repot"
+author: <a href="https://github.com/SimonDanisch">Simon Danisch</a>
+---
+
+A lot has been happening lately in Julia's GPU ecosystem.
+
+[Tim Maleadt](https://github.com/maleadt) has put great effort into making [CUDAnative.jl](https://github.com/JuliaGPU/CUDAnative.jl/) a fully fledged compiler backend to execute Julia code on nvidia GPUs.
```
Comment (maleadt) [inline blog/_posts/2017-8-25-gpuarrays.md]: That's not my name...

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+[Mike Innes](https://github.com/MikeInnes) just published [CuArrays](https://github.com/FluxML/CuArrays.jl) along with a detailed [blog post](http://mikeinnes.github.io/2017/08/24/cudanative.html) outlining all the great features Julia brings into the world of GPU programming.
+
+I want to add some more words about my recent work on [GPUArrays](https://github.com/JuliaGPU/GPUArrays.jl) and benchmarking our current GPU packages.
+
+GPUArrays is a backend independent N dimensional Array package that builds upon the work of [Transpiler.jl](https://github.com/SimonDanisch/Transpiler.jl) for an OpenGL and OpenCL backend and uses [CUDAnative](https://github.com/JuliaGPU/CUDAnative.jl/) for nvidia based hardware.
+It also offers a threaded Julia backend which can even utilize Xeon Phis and is great for debugging the GPU kernels.
+GPUArrays also includes abstractions over the intrinsics of the different backends and makes it possible to write GPU programs
+completely in Julia for all hardware.
+
+CuArrays is very similar to GPUArrays cudanative backend and was first supposed to be an extension of that,
```
Comment (maleadt) [inline blog/_posts/2017-8-25-gpuarrays.md]: > GPUArrays' CUDAnative backend

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+but we decided to concentrate on CUDA support in it's own package to create a much leaner package.
```
Comment (maleadt) [inline blog/_posts/2017-8-25-gpuarrays.md]: its

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+To get an idea of where we are in terms of performance, I started writing a benchmark suite for Julia's GPU array packages including [ArrayFire](https://github.com/gaika/ArrayFire.jl).
+It can run the whole suite and print out reports automatically. We're planning to integrate it into Github and publish the results on an interactive website where one can explore the results. Until then, the static report [over at GPUBenchmarks.jl](https://github.com/JuliaGPU/GPUBenchmarks.jl/blob/master/results/results.md) has to do the trick.
+This is the first release of GPUBenchmarks, so please treat the numbers with care!
+
+For the future, we plan to continue integrating existing libraries like [CUDNN](https://github.com/JuliaGPU/CUDNN.jl) and [CUSolver](https://github.com/JuliaGPU/CUSOLVER.jl) into GPUArrays as we head for feature parity with Julia's Base Array type.
+We also hope to make the ecosystem much more stable and improve the abstractions needed to write platform independent GPU code.
+
+All in all, Julia can now finally offer single source GPU programming, much like what [SyCL](https://www.khronos.org/sycl) has been promising for years.
```
Comment (maleadt) [inline blog/_posts/2017-8-25-gpuarrays.md]: Sounds pretty passive-aggressive IMO.

**Comment by maleadt** [comment]
Quick review. Also, because many of these packages are pretty hard to install, maybe add them to the Docker image I've been working on for users to quickly reproduce the benchmarks?

**Comment by SimonDanisch** [comment]
@ViralBShah, I linked the benchmarks, you think i should make it more prominent?!

**Comment by SimonDanisch** [comment]
@maleadt sure, what would be needed for that?

File: blog/_posts/2017-8-25-gpuarrays.md (hunk 4 above)
Comment (SimonDanisch) [reply blog/_posts/2017-8-25-gpuarrays.md]: haha, jeez, your Github name has been tricking me for ages :D

**Comment by maleadt** [comment]
I tagged you in the PR, right? It needs to pass CI on the tagged versions, after which the image will be automatically pushed.

**Comment by SimonDanisch** [comment]
So I just need to tag a working version? Somehow my tags havent been merged yet to METADATA

**Comment by SimonDanisch** [comment]
Ah nevermind, the tag PRs just started moving!

Am 27.08.2017 17:50 schrieb "Tim Besard" <notifications@github.com>:

**Comment by ViralBShah** [comment]
@simondanisch Almost nobody will click through. I would include the benchmarks in the post so that it is directly readable by anyone who reads the blog.

**Comment by ViralBShah** [comment]
Also the title is a bit generic. Could be specific to a particular milestone the GPU ecosystem has reached now.  Basically give me a specific reason to click on the tweet/RSS feed mention/etc for the full story.

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+---
+layout: post
+title: "Julia's GPU ecosystem: A status repot"
```
Comment (KristofferC) [inline blog/_posts/2017-8-25-gpuarrays.md]: report?

File: blog/_posts/2017-8-25-gpuarrays.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+but we decided to concentrate on CUDA support in its own package to create a much leaner package.
+This is a fair approach for now and we plan to merge things back at some point to decrease fragmentation of the ecosystem.
+
+Like CuArrays, GPUArrays supports functions like [broadcast](https://julialang.org/blog/2017/01/moredots), `mapreduce`, indexing, `vcat`/`hcat` and all the functionality that comes for free from the support of these generic functions, e.g. `sum`, `maximum`, etc.
+
+On top of that GPUArrays wraps CU/CL-BLAS and CU/CL-FFT enabling a wide range of applications.
+
+To get an idea of where we are in terms of performance, I started writing a benchmark suite for Julia's GPU array packages including [ArrayFire](https://github.com/gaika/ArrayFire.jl).
+It can run the whole suite and print out reports automatically. We're planning to integrate it into Github and publish the results on an interactive website where one can explore the results. Until then, the static report [over at GPUBenchmarks.jl](https://github.com/JuliaGPU/GPUBenchmarks.jl/blob/master/results/results.md) has to do the trick.
```
Comment (KristofferC) [inline blog/_posts/2017-8-25-gpuarrays.md]: "has to do the trick" sounds odd here, or?

**Comment by ViralBShah** [comment]
Bump. Are we publishing this? I think the benchmarks should be inline if we want to.

**Comment by SimonDanisch** [comment]
Yes, lets get this started again!

Am 28.10.2017 23:53 schrieb "Viral B. Shah" <notifications@github.com>:

**Comment by ViralBShah** [comment]
I believe this is already published.

**Comment by ViralBShah** [comment]
https://medium.com/@sdanisch/writing-extendable-and-hardware-agnostic-gpu-libraries-b21c145a8dad

**Pull request closed without merging**